core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.rpc;
+
+import org.apache.accumulo.core.trace.TraceUtil;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Scope;
+
+public class AccumuloProtocolFactory extends TCompactProtocol.Factory {
+
+  private static final long serialVersionUID = 1L;
+
+  private final boolean isClient;
+
+  public static class AccumuloProtocol extends TCompactProtocol {
+
+    private static final int MAGIC_NUMBER = 0x41434355; // "ACCU" in ASCII
+    private static final byte PROTOCOL_VERSION = 1;
+
+    private final boolean isClient;
+
+    private Span span = null;
+    private Scope scope = null;
+
+    public AccumuloProtocol(TTransport transport, boolean isClient) {
+      super(transport);
+      this.isClient = isClient;
+    }
+
+    @Override
+    public void writeMessageBegin(TMessage message) throws TException {
+      if (this.isClient) {
+        span = TraceUtil.startClientRpcSpan(this.getClass(), message.name);
+        scope = span.makeCurrent();
+
+        try {
+          this.writeHeader();
+        } catch (TException e) {
+          if (scope != null) {
```

Comment:
span has a recordException message, but not sure if that should be used here.  I have seen it but have not looked into it in detail.  Might be worth looking into.

Reply:
I took a look at the [Exceptions](https://opentelemetry.io/docs/specs/otel/trace/exceptions/#recording-an-exception) and [Recording errors](https://opentelemetry.io/docs/specs/semconv/general/recording-errors/) pages on the open telemetry docs and pushed commit 5dfab38 with what seems like the correct approach.